### DISCUSSION TITLE
Material Canvas generated shaders lack TransformColor

### DISCUSSION THREAD

**Discussion by ohmaya** [o3de/o3de#16407] [Help Wanted] [answered]
Connect a Texture node to Base Color, and do the same in Material Editor, the resulting materials are different in color.

In the generated <MaterialGraphName>_SurfaceEval.azsl, it is possible to add the following after the Sample() line (under "// O3DE_GENERATED_INSTRUCTIONS_BEGIN"):
`node11_sample_texture_2d_outColor.rgb = TransformColor(node11_sample_texture_2d_outColor.rgb, ColorSpaceId::LinearSRGB, ColorSpaceId::ACEScg);`
Then the colors are basically (but not entirely) the same.

There is the MaterialGraphName_SurfaceEval.azsl, under "// O3DE_GENERATED_INSTRUCTIONS_BEGIN", there is no "Sample". **How do I change some source files to let the shader generation process insert a "TransformColor" in the generated file?**

Thanks.

**Comment by ohmaya** [comment]
May be fixed by inserting TransformColor before:
`surface.SetAlbedoAndSpecularF0(inBaseColor, inSpecularF0Factor, inMetallic);`

But this is **added after all graph nodes and right before Base Color input property**. While the shader code for Material Editor only TransformColor for texture instead for color swatch.

**Comment by ohmaya** [comment]
Looks handled in MaterialGraphCompiler.
```
                // Gather and perform substitutions on instructions embedded directly in the node.
                AZStd::vector<AZStd::string> instructionsForNode;
                AtomToolsFramework::CollectDynamicNodeSettings(nodeConfig.m_settings, "instructions", instructionsForNode);
                AtomToolsFramework::ReplaceSymbolsInContainer(substitutionSymbols, instructionsForNode);
```

**Comment by galibzon** [comment]
@gadams3 ?

**Comment by gadams3** [comment, marked as answer]
I think the most flexible option would be adding a transform color node. This would allow uses to insert it anywhere in the graph and set or connect arbitrary inputs from the graph.

**Reply by ohmaya** [reply]
Or adding an DoTransformColor option to Texture nodes and generate the code in CollectDynamicNodeSettings letting TransformColor happen immediately after Sample.
